File: app/src/main/java/igrek/todotree/service/filesystem/FilesystemService.java
package igrek.todotree.service.filesystem;

import android.app.Activity;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.io.RandomAccessFile;
import java.util.ArrayList;
import java.util.List;

import igrek.todotree.logger.Logger;

public class FilesystemService {
	
	private Logger logger;
	private Activity activity;
	private ExternalCardService externalCardService;
	
	public FilesystemService(Logger logger, Activity activity, ExternalCardService externalCardService) {
		this.logger = logger;
		this.activity = activity;
		this.externalCardService = externalCardService;
	}
	
	public PathBuilder externalSDPath() {
		// returns internal dir but creates also /storage/extSdCard/Android/data/pkg - WTF?!
		//		activity.getExternalFilesDir("data");
		return new PathBuilder(externalCardService.getExternalSDPath());
	}
	
	public boolean mkdirIfNotExist(String path) {
		File f = new File(path);
		return !f.exists() && f.mkdirs();
	}
	
	private List<String> listDir(String path) {
		List<String> lista = new ArrayList<>();
		File f = new File(path);
		File file[] = f.listFiles();
		for (File aFile : file) {
			lista.add(aFile.getName());
		}
		return lista;
	}
	
	public List<String> listDir(PathBuilder path) {
		return listDir(path.toString());
	}
	
	private byte[] openFile(String filename) throws IOException {
		RandomAccessFile f = new RandomAccessFile(new File(filename), "r");
		int length = (int) f.length();
		byte[] data = new byte[length];
		f.readFully(data);
		f.close();
		return data;
	}
	
	public String openFileString(String filename) throws IOException {
		byte[] bytes = openFile(filename);
		return new String(bytes, "UTF-8");
	}
	
	private void saveFile(String filename, byte[] data) throws IOException {
		File file = new File(filename);
		FileOutputStream fos;
		fos = new FileOutputStream(file);
		fos.write(data);
		fos.flush();
		fos.close();
	}
	
	public void saveFile(String filename, String str) throws IOException {
		saveFile(filename, str.getBytes());
	}
	
	public boolean exists(String path) {
		File f = new File(path);
		return f.exists();
	}
	
	private boolean delete(String path) {
		File file = new File(path);
		return file.delete();
	}
	
	public boolean delete(PathBuilder path) {
		return delete(path.toString());
	}
	
	public void copy(File source, File dest) throws IOException {
		InputStream is = null;
		OutputStream os = null;
		try {
			is = new FileInputStream(source);
			os = new FileOutputStream(dest);
			byte[] buffer = new byte[1024];
			int length;
			while ((length = is.read(buffer)) > 0) {
				os.write(buffer, 0, length);
			}
		} finally {
			if (is != null) {
				is.close();
			}
			if (os != null) {
				os.close();
			}
		}
	}
}

File: app/src/main/java/igrek/todotree/service/filesystem/FilesystemService.kt
package igrek.todotree.service.filesystem

import android.content.Context
import android.os.Environment
import igrek.todotree.info.logger.Logger
import igrek.todotree.info.logger.LoggerFactory
import igrek.todotree.inject.LazyExtractor
import igrek.todotree.inject.LazyInject
import igrek.todotree.inject.appFactory
import java.io.*
import java.nio.charset.Charset

open class FilesystemService(
    context: LazyInject<Context> = appFactory.context,
) {
    val context by LazyExtractor(context)

    private val logger: Logger = LoggerFactory.logger

    fun appDataRootDir(): File {
        context.filesDir.takeIf { it.isDirectory }?.let { return it }
        return File("/data/data/${context.packageName}/files").createIfMissing()
    }

    fun externalAppDataRootDir(): File {
        return context.getExternalFilesDir("files")!!.createIfMissing()
    }

    fun internalStorageAppDir(): File {
        val internalStorageDir = Environment.getExternalStorageDirectory()
        val dir = internalStorageDir.resolve(".todotree")
        dir.mkdirs()
        return dir
    }

    private fun File.createIfMissing(): File {
        return this.also {
            it.takeIf { !it.exists() }?.let { file ->
                file.mkdir()
                logger.info("missing directory created: ${file.absolutePath}")
            }
        }
    }

    fun appDataSubDir(name: String): File {
        return appDataRootDir().resolve(name).createIfMissing()
    }

    fun listDirFilenames(dir: File): List<String> {
        return dir.listFiles().map { it.name }
    }

    @Throws(IOException::class)
    private fun openFile(filename: String): ByteArray {
        val f = RandomAccessFile(File(filename), "r")
        val length = f.length().toInt()
        val data = ByteArray(length)
        f.readFully(data)
        f.close()
        return data
    }

    @Throws(IOException::class)
    fun openFileString(filename: String): String {
        val bytes = openFile(filename)
        return String(bytes, Charset.forName("UTF-8"))
    }

    @Throws(IOException::class)
    private fun saveFile(filename: String, data: ByteArray) {
        val file = File(filename)
        createMissingParentDir(file)
        val fos = FileOutputStream(file)
        fos.write(data)
        fos.flush()
        fos.close()
    }

    private fun createMissingParentDir(file: File) {
        val parentDir = file.parentFile
        if (!parentDir.exists()) {
            parentDir.mkdir()
            logger.debug("missing dir created: $parentDir")
        }
    }

    @Throws(IOException::class)
    fun saveFile(filename: String, str: String) {
        saveFile(filename, str.toByteArray())
    }

    @Throws(IOException::class)
    fun copy(source: File?, dest: File?) {
        var `is`: InputStream? = null
        var os: OutputStream? = null
        try {
            `is` = FileInputStream(source)
            os = FileOutputStream(dest)
            val buffer = ByteArray(1024)
            var length: Int
            while (`is`.read(buffer).also { length = it } > 0) {
                os.write(buffer, 0, length)
            }
        } finally {
            `is`?.close()
            os?.close()
        }
    }

}